pybossa/sched.py

```diff
@@ -590,20 +591,31 @@ def acquire_lock(task_id, user_id, limit, timeout, pipeline=None, execute=True):
     return False
 
 
-def release_reserve_task_lock_by_id(project_id, task_id, user_id, timeout, expiry=EXPIRE_RESERVE_TASK_LOCK_DELAY):
+def release_reserve_task_lock_by_id(project_id, task_id, user_id, timeout, expiry=EXPIRE_RESERVE_TASK_LOCK_DELAY, release_all_task=None):
     reserve_key = get_reserve_task_key(task_id)
     if not reserve_key:
         return
 
     redis_conn = sentinel.master
     lock_manager = LockManager(redis_conn, timeout)
-    resource_id = "reserve_task:project:{}:category:{}:user:{}:task:{}".format(
-        project_id, reserve_key, user_id, task_id)
-    lock_manager.release_reserve_task_lock(resource_id, expiry)
-    current_app.logger.info(
-        "Release reserve task lock. project %s, task %s, user %s, expiry %d",
-        project_id, task_id, user_id, expiry
-    )
+    if release_all_task:
+        pattern = "reserve_task:project:{}:category:{}:user:{}:task:*".format(
+            project_id, reserve_key, user_id)
+        resource_ids = lock_manager.scan_keys(pattern)
+        for k in resource_ids:
+            task_id_in_key = int(k.decode().split(":")[-1])
+            # If a task is locked by the user(in other tab), then the category lock should not be released
+            if task_id_in_key == task_id or not lock_manager.has_lock(get_task_users_key(task_id_in_key), user_id):
+                lock_manager.release_reserve_task_lock(k, expiry)
```

Comment:
Question: the `EXPIRE_RESERVE_TASK_LOCK_DELAY` is 30min, it makes sense to have 30min expiration delay when user clicks "submit", however, is it expected to make the delay time 30min in the case that user wants to "cancel" or "submit and leave"?

Reply:
When "cancel" or "submit and leave" actions calls ```release_reserve_task_lock_by_id```, it pass ```EXPIRE_LOCK_DELAY``` to it, which is 5 seconds. It won't get 30min (the default value) unless you omit it. But we don't want to hard coded to ```EXPIRE_LOCK_DELAY``` in the implementation, so that we can have a little more flexibility.

Reply:
Passing 'EXPIRE_LOCK_DELAY` makes sense to me, thanks you.